src/XHRConnection.js

```diff
@@ -127,80 +156,111 @@ export default class XHRConnection extends AbstractConnection {
 
   /**
    * create, prepare, open and send the xhr request
+   * @param {array} - header.
+   * @return {XHRconnection} - this connection
    */
-  open() {
+  open(headers) {
     if (this.state !== XHRConnection.INIT) {
       return;
     }
 
     this.xhr.addEventListener("progress", () => {
-      this.emit(
-        ConnectionEvent.DATA,
-        new ConnectionEvent(this, ConnectionEvent.DATA)
-      );
+      window.setTimeout(() => {
+        this.emit(
+          ConnectionEvent.DATA,
+          new ConnectionEvent(this, ConnectionEvent.DATA)
+        );
+      }, 0);
     });
 
     this.xhr.addEventListener("load", () => {
       this.state = XHRConnection.CLOSED;
+      this.closed = Date.now();
 
       if (this.status < 400) {
-        this.emit(
-          ConnectionEvent.COMPLETE,
-          new ConnectionEvent(this, ConnectionEvent.COMPLETE)
-        );
+        window.setTimeout(() => {
+          this.emit(
+            ConnectionEvent.COMPLETE,
+            new ConnectionEvent(this, ConnectionEvent.COMPLETE)
+          );
+        }, 0);
 
         return;
       }
 
-      this.emit(
-        ConnectionEvent.ERROR,
-        new ConnectionEvent(this, ConnectionEvent.ERROR)
-      );
+      window.setTimeout(() => {
+        this.emit(
+          ConnectionEvent.ERROR,
+          new ConnectionEvent(this, ConnectionEvent.ERROR)
+        );
+      }, 0);
     });
 
     this.xhr.addEventListener("abort", () => {
       this.state = XHRConnection.CLOSED;
+      this.closed = Date.now();
 
-      this.emit(
-        ConnectionEvent.ABORT,
-        new ConnectionEvent(this, ConnectionEvent.ABORT)
-      );
+      window.setTimeout(() => {
+        this.emit(
+          ConnectionEvent.ABORT,
+          new ConnectionEvent(this, ConnectionEvent.ABORT)
+        );
+      }, 0);
     });
 
     this.xhr.addEventListener("error", () => {
       this.state = XHRConnection.CLOSED;
+      this.closed = Date.now();
 
-      this.emit(
-        ConnectionEvent.ERROR,
-        new ConnectionEvent(this, ConnectionEvent.ERROR)
-      );
+      window.setTimeout(() => {
+        this.emit(
+          ConnectionEvent.ERROR,
+          new ConnectionEvent(this, ConnectionEvent.ERROR)
+        );
+      }, 0);
     });
 
     this.xhr.addEventListener("timeout", () => {
       this.state = XHRConnection.CLOSED;
+      this.closed = Date.now();
 
-      this.emit(
-        ConnectionEvent.ERROR,
-        new ConnectionEvent(this, ConnectionEvent.ERROR)
-      );
+      window.setTimeout(() => {
+        this.emit(
+          ConnectionEvent.ERROR,
+          new ConnectionEvent(this, ConnectionEvent.ERROR)
+        );
+      }, 0);
     });
 
     this.xhr.open(this.method, this.url);
     this.xhr.timeout = this.timeout;
     this.state = XHRConnection.OPEN;
-    this.emit(
-      ConnectionEvent.OPEN,
-      new ConnectionEvent(this, ConnectionEvent.OPEN)
-    );
+    this.opened = Date.now();
+    window.setTimeout(() => {
```

Comment:
Can we move the setTimeout to the callee side (so `emit`)? It seems pretty repetitive, like a good default and easy to forget